qubes/storage/__init__.py

```diff
@@ -158,6 +158,48 @@ def __xml__(self):
         config = _sanitize_config(self.config)
         return lxml.etree.Element('volume', **config)
 
+    def ephemeral(self):
+        '''Should this volume be encrypted with an ephemeral key in dom0?
+        '''
+        return not self.snap_on_start and not self.save_on_stop and \
+                self.domain is None and self.rw
+
+    async def start_encrypted(self, name):
+        '''
+        Start a volume encrypted with an ephemeral key.
+        This can be implemented as a coroutine.
+
+        The default implementation of this method uses ``cryptsetup(8)`` with a
+        key taken from ``/dev/urandom``.  This is highly secure and works with
+        any storage pool implementation.  Volume implementations should override
+        this method if they can provide a secure and more efficient
+        implementation.
+        '''
+        if self.script is not None:
+            self._not_implemented("starting an encrypted volume with a script")
+        await qubes.utils.coro_maybe(self.start())
+        await qubes.utils.cryptsetup(
+            '--key-file=/dev/urandom',
+            '--cipher=aes-xts-plain64',
+            '--type=plain',
+            '--',
+            'open',
+            self.path,
```

Comment:
`self.path` may be not the final device for the VM to use, you need to extract it from `self.block_device()` (after `self.start()`). Coincidentally for all the pools currently in core-admin, both paths are the same for volatile volumes (but for example they do differ in some other cases).